File: backtracking/combination_sum_iv.rb
# Approach 1: Recursion with memoization
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def combination_sum4(nums, target)
    return 0 if !nums || nums.empty?
    recurse(nums, target, {0 => 1})
end

def recurse(nums, target, cache)
    return cache[target] if cache.key?(target)
    cache[target] = 0 if !cache.key?(target)

    nums.each do |x|
        next if target - x < 0
        cache[target] += recurse(nums, target - x, cache)
    end

    cache[target]
end

# Approach 2: Bottom-up DP
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def combination_sum4(nums, target)
    dp = Array.new(target + 1, 0)
    dp[0] = 1

    1.upto(target) do |i|
        nums.each do |x|
            dp[i] += dp[i - x] if i - x >= 0
        end
    end

    dp[target]
end

# Combination Sum IV
# https://leetcode.com/problems/combination-sum-iv/

# Approach 1: Dynamic programming top-down (Recursion with memoization)
# Approach 2: DP bottom-up

# Follow Up
# What if negative numbers are allowed?
# PROBLEM: For negative numbers the combination length can be unbounded,
# as we can keep subtracting and adding 1 unlimited number of times to get the same target
# SOLUTION: Keep a bound k on the combination sequence length.

# Most similar to combination sum ii as duplicates are allowed
# Similar to word_break II in terms of recurse with memoization

File: backtracking/combination_sum_iv.py
from typing import List


# Approach 1: Top Down DP a.k.a Memoization
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        self.cache = {0:1}
        return self.dfs(nums, target, )

    def dfs(self, nums, target):
        if target in self.cache: return self.cache[target]

        self.cache[target] = 0
        for x in nums:
            if target - x < 0: continue
            self.cache[target] += self.dfs(nums, target - x)

        return self.cache[target]



# Approach 2: Bottom-Up DP a.k.a Tabulation
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        results = [0] * (target + 1)
        results[0] = 1
        for i in range(1, target + 1):
            for x in nums:
                if i - x < 0: continue
                results[i] += results[i - x]

        return results[target]



# Combination Sum IV 
# https://leetcode.com/problems/combination-sum-iv/

# Approach 1: Top Down DP a.k.a Memoization
# Time: O(target * nums)
# Space: O(target * nums)

# Approach 2: Bottom-Up DP a.k.a Tabulation


# Follow Up
# What if negative numbers are allowed?
# We need additional constraints in the problem else we will have an infinite loop.
# https://discuss.leetcode.com/topic/53553/what-if-negative-numbers-are-allowed-in-the-given-array


# ALL combination sum problems
# https://leetcode.com/problems/combination-sum-iv/discuss/85120/C++-template-for-ALL-Combination-Problem-Set

# https://leetcode.com/problems/combination-sum-iv/discuss/85106/A-summary-of-all-combination-sum-problem-in-LC-C++
